### ISSUE TITLE
Split up tests?

### ISSUE THREAD

**Issue by max-sixty** [pydata/xarray#5409]
Currently a large share of our tests are in `test_dataset.py` and `test_dataarray.py` — each of which are around 7k lines.

There's a case for splitting these up:
- Many of the tests are somewhat duplicated between the files (and `test_variable.py` in some cases) — i.e. we're running the same test over a Dataset & DataArray, but putting them far away from each other in separate files. Should we instead have them split by "function"; e.g. `test_rolling.py` for all rolling tests?
- My editor's takes 5-20 seconds to run the linter and save the file. This is a very narrow complaint.
- Now that we're all onto pytest, there's no need to have them in the same class.

If we do this, we could start on the margin — new tests around some specific functionality — e.g. join / rolling / reindex / stack (just a few from browsing through) — could go into a new respective `test_{}.py` file. Rather than some big copy and paste commit.

**Comment by keewis** [comment]
that sounds like a good idea, and we'll definitely have to do this gradually because PRs with several thousand lines of changes are really difficult to review.

Probably unrelated, but I'd also like to refactor / reorder the tests in `test_combine.py`.

**Comment by shoyer** [comment]
Yes, splitting up these giant test files would definitely be a good idea. `test_backends.py` might be another good candidate.

I don't think there was ever a good reason why they _needed_ be in a single class. `unittest` works just fine with multiple test classes, too.

**Comment by max-sixty** [comment]
Great, it seems there's some backing for this. What's the best way to start?

I'm happy to do a PR to pull all the `rolling` tests into `test_rolling.py` as an example to start. I'm also happy to say any _new_ rolling tests need to go into `test_rolling.py`, to minimize movement, as suggested above. But maybe that's hard to manage, and it's not obvious that's the policy for new contributors.

**Comment by TomNicholas** [comment]
This is kind of a general question whenever we add new features: if I add to the API such that I can access my new computation `newfunc` via any of `da.newfunc()`, `ds.newfunc()` or `xr.newfunc()`, then should the tests live in `test_dataarray.py`, `test_dataset.py`, `test_computation.py`, all three of them, or make a new file `test_newfunc.py`? I agree we should move towards the latter, and try to just make `test_dataset.py` & `test_dataset.py` test the core functionality of those objects, and not their extensive method APIs.
